Update launch error handling

ninaniiranen opened the pull request:
This will allow SDK users to be aware and react to errors when dreams is launched.

When launching dreams a function can be passed that will be invoked with the result of the launch.
This `Result<Unit, LaunchError>` will tell the user if the call failed or succeeded.
If it failed the `LaunchError` will contain information on why.

`Location` has been removed as it's unused.

Commit:
refactor: rewrite result class

This exposes the result class and makes it a bit more flexible.

Commit:
feat: allow callers of launch to react to errors

Adds a functional interface as a parameter to launch. This SAM is invoked with
the result of the launch call. If the launch fails, the result object will
contain information as to why it failed. If the result is successful there is
no extra data.

The default SAM will just log errors to logcat.

Commit:
refactor: removed unused parameter from launch

BREAKING CHANGE: `DreamsView.launch` no longer takes a location.

Commit:
chore: update example app to use launch result

Commit:
build: update dependencies

ochococo commented:
📃  I think README should be updated as well, will review the rest tomorrow first thing.

Commit:
docs: update documentation

Commit:
test: add launch error tests

Commit:
test: fixed flaky test

In some cases the launch completion function was not called in time.

Commit:
test: fix tests failing on API 21

The javascript in index.html was not valid using such an old web view
implementation.
Mockk seems to have some issues with spying on functional interfaces, see:
https://github.com/mockk/mockk/issues/258

Commit:
test: don't crash mock server on missing token
